modules/apps/object-forms/object-forms-storage/src/main/java/com/liferay/object/forms/storage/internal/storage/ObjectsDDMStorageAdapter.java

```diff
@@ -0,0 +1,208 @@
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.object.forms.storage.internal.storage;
+
+import com.liferay.dynamic.data.mapping.exception.StorageException;
+import com.liferay.dynamic.data.mapping.model.DDMForm;
+import com.liferay.dynamic.data.mapping.model.DDMFormField;
+import com.liferay.dynamic.data.mapping.model.LocalizedValue;
+import com.liferay.dynamic.data.mapping.model.Value;
+import com.liferay.dynamic.data.mapping.service.DDMFieldLocalService;
+import com.liferay.dynamic.data.mapping.storage.DDMFormFieldValue;
+import com.liferay.dynamic.data.mapping.storage.DDMFormValues;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapter;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterDeleteRequest;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterDeleteResponse;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterGetRequest;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterGetResponse;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterSaveRequest;
+import com.liferay.dynamic.data.mapping.storage.DDMStorageAdapterSaveResponse;
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectEntry;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectEntryLocalService;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.Serializable;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Gabriel Albuquerque
+ */
+@Component(
+	immediate = true, property = "ddm.storage.adapter.type=objects",
+	service = DDMStorageAdapter.class
+)
+public class ObjectsDDMStorageAdapter implements DDMStorageAdapter {
+
+	@Override
+	public DDMStorageAdapterDeleteResponse delete(
+			DDMStorageAdapterDeleteRequest ddmStorageAdapterDeleteRequest)
+		throws StorageException {
+
+		ObjectEntry objectEntry;
+
+		try {
+			objectEntry = _objectEntryLocalService.getObjectEntry(
+				ddmStorageAdapterDeleteRequest.getPrimaryKey());
+
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			_ddmFieldLocalService.deleteDDMFormValues(
+				(Long)values.get("ddmStorageId"));
+
+			_objectEntryLocalService.deleteObjectEntry(
+				objectEntry.getObjectEntryId());
```

Comment:
These instruction should be inside a transaction, or you should simulate it. If it raises an error on line 77, the DDMFormValue is gone, but the related ObjectEntry is still there. It shouldn't occurs.

Reply:
I think when we have a StorageException, the operation rollbacks the changes. But I can test it, of course.